pkg/datarepair/checker/online.go

```diff
@@ -0,0 +1,72 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package checker
+
+import (
+	"context"
+	"time"
+
+	"storj.io/storj/pkg/overlay"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+)
+
+// ReliableCache caches the reliable nodes for the specified staleness duration
+// and updates automatically from overlay.
+type ReliableCache struct {
+	overlay    *overlay.Cache
+	staleness  time.Duration
+	lastUpdate time.Time
+	reliable   map[storj.NodeID]struct{}
+}
+
+// NewReliableCache creates a new reliability checking cache.
+func NewReliableCache(overlay *overlay.Cache, staleness time.Duration) *ReliableCache {
+	return &ReliableCache{
+		overlay:   overlay,
+		staleness: staleness,
+		reliable:  map[storj.NodeID]struct{}{},
+	}
+}
+
+// LastUpdate returns when the cache was last updated.
+func (cache *ReliableCache) LastUpdate() time.Time { return cache.lastUpdate }
+
+// MissingPieces returns piece indices that are unreliable with the given staleness period.
+func (cache *ReliableCache) MissingPieces(ctx context.Context, created time.Time, pieces []*pb.RemotePiece) ([]int32, error) {
+	if created.After(cache.lastUpdate) || time.Since(cache.lastUpdate) > cache.staleness {
+		err := cache.Refresh(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var unreliable []int32
+	for _, piece := range pieces {
+		if _, ok := cache.reliable[piece.NodeId]; !ok {
+			unreliable = append(unreliable, piece.PieceNum)
+		}
+	}
+	return unreliable, nil
+}
+
+// Refresh refreshes the cache.
+func (cache *ReliableCache) Refresh(ctx context.Context) error {
+	for id := range cache.reliable {
+		delete(cache.reliable, id)
```

Comment:
This surprised me. Is this an idiom for Go mutexes? I admit I don't know. Wouldn't it be a lot faster just to reinstantiate the map entirely?

Or, if not, do you not need some sort of mutex around this? Is it ok for some other goroutine to query missing pieces while half of the items are deleted from the cache? Or maybe only one goroutine will ever use it? Either way, it is probably worth leaving a comment to avoid confusing the future reader.

Reply:
This deletion loop gets optimized away to doing the most efficient clearing.

The implementation assumes that it's used from one goroutine only. I guess we could add a mutex around it without problems.

Reply:
I didn't know that about the optimization! That is cool.

I would be happy with just a comment somewhere saying that `ReliabilityCache` can only be used from a single goroutine.

Reply:
added comment that `ReliabilityCache` is not safe for concurrent use